library/modules/EventManager.cpp

```diff
@@ -193,1087 +240,1564 @@ std::array<eventManager_t,EventType::EVENT_MAX> compileManagerArray() {
     return managers;
 }
 
+template<typename T, bool has_duplicates = false>
+class event_tracker {
+private:
+    std::unordered_map<T, int32_t> seen;
+    std::multimap<int32_t, T> history;
+public:
```

Comment:
Sending events on multiple tick frequencies will require a separate PR as well, as it changes the code most substantially, and myk requested a full writeup.. both of which warrant a completely separate PR for these changes.

Reply:
Specifically, what I'm looking for is justification for the change. I want to see how the added complexity makes real world use cases possible. How would current users and actual planned future users benefit?

The usual behavior of an event trigger framework is to notify the listeners when an event occurs. Our current implementation is closer to this behavior than the proposed one, which is more of a watermarked pub/sub architecture. Pub/sub can certainly be useful, but it's not obvious to me how it's useful for DFHack's use case.